sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java

```diff
@@ -0,0 +1,143 @@
+
+  private static final Logger LOG = Loggers.get(ContextualServer.class);
+
+  private Configuration configuration;
+  private ExecutableBundleFactory bundleFactory;
+  private static AtomicReference<Process> serverProcess = new AtomicReference<>();
+
+  public ContextualServer(ExecutableBundleFactory bundleFactory) {
+    this(null, bundleFactory);
+  }
+
+  public ContextualServer(Configuration configuration, ExecutableBundleFactory bundleFactory) {
+    this.configuration = configuration;
+    this.bundleFactory = bundleFactory;
+  }
+
+  @Override
+  public void start() {
+    LOG.info("Starting SonarTS Server");
+
+    if (serverProcess.get() != null && serverProcess.get().isAlive()) {
+      LOG.warn("Skipping SonarTS Server start, already running");
+      return;
+    }
+
+    if (configuration == null) {
+      LOG.warn("Skipping SonarTS Server start due to missing configuration");
+      return;
+    }
+    
+    startSonarTSServer();
+  }
+
+  private void startSonarTSServer() {
+    ExecutableBundle bundle = bundleFactory.createAndDeploy(getServerDir(), configuration);
+
+    ProcessBuilder processBuilder = new ProcessBuilder(bundle.getSonarTSServerCommand().commandLineTokens());
+    setNodePath(processBuilder);
+
+    try {
+      processBuilder.inheritIO();
+      serverProcess.set(processBuilder.start());
+      //TODO find a better way to wait until the server is up
+      // e.g. loop with accessing the port
+      synchronized (this) {
+        wait(1000);
+      }
+      LOG.info("SonarTS Server is started");
+
+    } catch (IOException e) {
+      LOG.error("Failed to start SonarTS Server", e);
+
+    } catch (InterruptedException e) {
+      LOG.error("Failed to wait until the SonarTS Server is up", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void setNodePath(ProcessBuilder processBuilder) {
+    Optional<String> typescriptLocation = configuration.get(TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY);
+
+    if (typescriptLocation.isPresent()) {
+      SensorContextUtils.setNodePath(new File(typescriptLocation.get()), processBuilder);
+
+    } else {
+      LOG.error("No value provided by SonarLint for TypeScript location; property " + TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY + " is missing");
+    }
+  }
+
+  private static File getServerDir() {
+    File serverFolder = new File(System.getProperty("java.io.tmpdir"), "sonarts");
+    if (!serverFolder.exists()) {
+      serverFolder.mkdir();
+    }
+    return serverFolder;
+  }
+
+  @Override
+  public void stop() {
+    Process process = serverProcess.get();
+    if (process != null && process.isAlive()) {
+
+      process.destroy();
+      synchronized (this) {
+        try {
+          wait(200);
+        } catch (InterruptedException e) {
+          LOG.error("Failed to wait until the SonarTS Server is stopped", e);
+          throw new RuntimeException(e);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define and throw a dedicated exception instead of using a generic one. ([squid:S00112](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00112))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-351&issues=AWD5U5d9EFfjxkQcPdcf&open=AWD5U5d9EFfjxkQcPdcf)

[ISSUE_KEY]: AWD5U5d9EFfjxkQcPdcf